source/extensions/tracers/datadog/agent_http_client.cc

```diff
@@ -0,0 +1,140 @@
+#include "source/extensions/tracers/datadog/agent_http_client.h"
+
+#include <datadog/dict_reader.h>
+#include <datadog/dict_writer.h>
+#include <datadog/error.h>
+
+#include <cassert>
+#include <cstddef>
+#include <cstdint>
+#include <datadog/json.hpp>
+#include <memory>
+#include <string>
+
+#include "source/common/http/message_impl.h"
+#include "source/common/http/utility.h"
+#include "source/extensions/tracers/datadog/dict_util.h"
+#include "source/extensions/tracers/datadog/string_util.h"
+#include "source/extensions/tracers/datadog/tracer_stats.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Tracers {
+namespace Datadog {
+
+AgentHTTPClient::AgentHTTPClient(Upstream::ClusterManager& cluster_manager,
+                                 const std::string& cluster, const std::string& reference_host,
+                                 TracerStats& stats)
+    : collector_cluster_(cluster_manager, cluster), cluster_(cluster),
+      reference_host_(reference_host), stats_(&stats) {}
+
+AgentHTTPClient::~AgentHTTPClient() {
+  for (const auto& [request_ptr, _] : handlers_) {
+    assert(request_ptr);
+    request_ptr->cancel();
+  }
+}
+
+// dd::HTTPClient
+
+dd::Expected<void> AgentHTTPClient::post(const URL& url, HeadersSetter set_headers,
+                                         std::string body, ResponseHandler on_response,
```

Comment:
`const std::string& body`

I'm not sure about the other by-value objects.

Reply:
This function is an override of `virtual ... datadog::tracing::HTTPClient::post(...)`: https://github.com/DataDog/dd-trace-cpp/blob/cdb5f235103f7733cb5b8353ac565be42e833af6/src/datadog/http_client.h#L40-L49

The signatures must match, so we can't change it here without also changing it everywhere else (in the base library and in all clients of the base library... which isn't all that many clients right now).

I chose `std::string body` instead of `const std::string& body` so that the value can be `move`d from its original source if possible.

As for the `std::function` callbacks, I'm not sure why I passed those by value and not by `const&`. I suppose the same `move` argument applies, but for the callbacks it's unlikely to matter.

Reply:
Cool. I had to do some experiments to see if that usage model actually saved the copy; it does appear to.